cpp/src/arrow/array/builder_base.h

```diff
@@ -116,6 +116,10 @@ class ARROW_EXPORT ArrayBuilder {
   /// This method is useful when appending null values to a parent nested type.
   virtual Status AppendEmptyValues(int64_t length) = 0;
 
+  /// \brief Append a value from a scalar
+  virtual Status AppendScalar(const Scalar& scalar);
+  virtual Status AppendScalars(const ScalarVector& scalars);
```

Comment:
I would also suggest the overload `AppendScalar(const Scalar& scalat, int64_t nrepeats)` to append the same scalar a number of times.

Reply:
Good point, I've added the overload.